services/community-metrics/src/updateBeneficiaries.ts

```diff
@@ -0,0 +1,99 @@
+import { database, utils, subgraph, config } from '@impactmarket/core';
+import { getAddress } from '@ethersproject/address';
+import { Transaction } from 'sequelize';
+
+export async function updateBeneficiaries(): Promise<void> {
+    utils.Logger.info('Updating beneficiaries...');
+    const t = await database.sequelize.transaction();
+
+    try {
+        const lastUpdate = await database.models.subgraphUBIBeneficiary.findOne({
+            attributes: ['updatedAt'],
+            order: [['updatedAt', 'DESC']]
+        });
+
+        const entityLastUpdated = lastUpdate ? (lastUpdate.updatedAt.getTime() / 1000) | 0 : undefined;
+
+        if (!entityLastUpdated) {
+            await updateAllBeneficiaries(t);
+        } else {
+            await updateBeneficiariesState(entityLastUpdated, t);
+        }
+        
+        await t.commit();
+        utils.Logger.info('Beneficiaries updated!');
+    } catch (error) {
+        await t.rollback();
+        utils.Logger.error('Error update beneficiaries: ', error);
+        utils.slack.sendSlackMessage('🚨 Error to update beneficiaries', config.slack.lambdaChannel);
+    }
+}
+
+async function updateBeneficiariesState(searchParam: number | string,  t: Transaction) {
+    const { subgraphUBIBeneficiary } = database.models;
+    const batchSize = config.cronJobBatchSize;
+    let offset = 0;
+
+    while (true) {
+        const result = await subgraph.queries.beneficiary.getBeneficiaryState(
+            `first: ${batchSize}
+            skip: ${offset}
+            ${typeof searchParam === 'number'
+                ? `where: { lastActivity_gt: ${searchParam} }`
+                : `where: { community:"${searchParam.toLowerCase()}" }`}`,
+            `
+                address
+                claims
+                community {
+                    id
+                }
+                lastClaimAt
+                preLastClaimAt
+                since
+                claimed
+                state
+            `
+        );
+
+        const beneficiaries = result || [];
+
+        if (beneficiaries.length === 0) {
+            break;
+        }
+
+        offset += batchSize;
+
+        await Promise.all(beneficiaries.map(beneficiary => {
+            return subgraphUBIBeneficiary
+                .findOne({ where: { userAddress: getAddress(beneficiary.address) } })
+                .then(model => {
+                    beneficiary['communityAddress'] = getAddress(beneficiary.community.id);
+                    beneficiary['userAddress'] = getAddress(beneficiary.address);
+    
+                    if(model) {
+                        return subgraphUBIBeneficiary.update(beneficiary as any, {
+                            where: { id: model.id },
+                            transaction: t,
+                        }) as any;
+                    }
+                    return subgraphUBIBeneficiary.create(beneficiary as any, { transaction: t });
```

Comment:
would it make sense to clear the beneficiaries' endpoints cache somewhere here?

Reply:
Yes, totally sense